### PR TITLE
ec2_instance - Support AdditionalInfo option.

### PR DESCRIPTION
**Pull request by yugesk**
This is a reserved parameter that was already supported by boto3.

##### SUMMARY
Enable support for passing additional_info while creating instance. 

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
ec2_instance

### CODE REVIEW

**Comment by tremble** [comment]
@yugesk,

Thanks for taking the time to submit this PR.  https://github.com/ansible-collections/amazon.aws/pull/1825 was recently submitted which would support a broader range of placement options.

I think it might be better to reduce this PR down to just adding the "additional_info" parameter.  IMO grouping the placement options (as with #1825) is going to be more sustainable (EC2 instances have a huge number of potential options)

**Comment by yugesk** [comment]
> @yugesk,
> 
> Thanks for taking the time to submit this PR. #1825 was recently submitted which would support a broader range of placement options.
> 
> I think it might be better to reduce this PR down to just adding the "additional_info" parameter. IMO grouping the placement options (as with #1825) is going to be more sustainable (EC2 instances have a huge number of potential options)
> 
> Please also add some integration tests (see other examples in tests/integration/targets/ec2_instance_*)

Thank you for taking a look at this PR, I will update to only add additional_info now that #1825 will take care of other things I need.

**Comment by yugesk** [comment]
@tremble AdditionalInfo is a property reserved for internal use and it will fail for external users I think. How do I add integration tests to it?

**Comment by yugesk** [comment]
Could someone please review this PR?

**Comment by yugesk** [comment]
@tremble could you please review again?

**Comment by tremble** [comment]
> @tremble AdditionalInfo is a property reserved for internal use and it will fail for external users I think. How do I add integration tests to it?

If it's reserved for internal use, what's the use case for adding it?

**Review by tremble — APPROVED** [review]
We also need a changelog fragment, but otherwise I'm ok with this change: it's reasonably minimal.

However, I'd like to hear from @alinabuzachis or @jillr how they feel about supporting something that (if I understand correctly) can only be used by Amazon Engineers.

**Comment by tremble** [comment]
( Changelog docs https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs-how-to )

**Comment by gravesm** [comment]
I'm a little confused what's going on with CI here, because this PR should fail sanity checks as is. It fails locally when I run it. We can't have an undocumented parameter. This is a pretty weird case. At the very least, the new parameter needs to be documented and the documentation should clearly state that it is unsupported and will likely fail. Probably something about it being added purely for Amazon's internal use.

**Comment by tremble** [comment]
> I'm a little confused what's going on with CI here, because this PR should fail sanity checks as is. It fails locally when I run it. We can't have an undocumented parameter.

Some of the tests need to be manually approved when a new contributor submits a PR.  This will be more important if/when integration tests move to github actions.

File: plugins/modules/ec2_instance.py
Comment (tremble) [inline plugins/modules/ec2_instance.py:2285]: The type listed in the arguments doesn't match the type listed in the return and parameters documentation.

https://github.com/ansible-collections/amazon.aws/actions/runs/6798846893/job/18485232245?pr=1828

```
ERROR: plugins/modules/ec2_instance.py:0:0: doc-type-does-not-match-spec: Argument 'additional_info' in argument_spec defines type as 'str' but documentation defines type as 'dict'
```

File: plugins/modules/ec2_instance.py
Comment (tremble) [inline plugins/modules/ec2_instance.py:378]:
```suggestion
      - Reserved for Amazon's internal use.
```

File: plugins/modules/ec2_instance.py
Comment (tremble) [inline plugins/modules/ec2_instance.py:956]:
```suggestion
            description: Reserved for Amazon's internal use.
```

https://github.com/ansible-collections/amazon.aws/actions/runs/6798846893/job/18485232245?pr=1828

```
ERROR: Found 2 pep8 issue(s) which need to be resolved:
ERROR: plugins/modules/ec2_instance.py:378:44: W291: trailing whitespace
ERROR: plugins/modules/ec2_instance.py:956:61: W291: trailing whitespace
```

**Review by tremble — CHANGES_REQUESTED** [review]

File: plugins/modules/ec2_instance.py
Comment (yugesk) [reply plugins/modules/ec2_instance.py]: Updated in latest rev.

File: plugins/modules/ec2_instance.py
Comment (yugesk) [reply plugins/modules/ec2_instance.py]: updated in latest rev.

File: plugins/modules/ec2_instance.py
Comment (yugesk) [reply plugins/modules/ec2_instance.py:2285]: fixed now.

File: plugins/modules/ec2_instance.py
Comment (tremble) [inline plugins/modules/ec2_instance.py:1449]:
```suggestion
    if params.get("additional_info"):
        spec["AdditionalInfo"] = params.get("additional_info")
```

**Review by tremble — APPROVED** [review]

File: plugins/modules/ec2_instance.py
Comment (tremble) [inline plugins/modules/ec2_instance.py:2285]:
```suggestion
        additional_info=dict(type="str"),
```

**Comment by yugesk** [comment]
@tremble thanks for fixing the black formatting errors. Why are still some tests failing?

**Comment by tremble** [comment]
recheck

**Comment by tremble** [comment]
> @tremble thanks for fixing the black formatting errors. Why are still some tests failing?

Some of the tests are slightly flakey.  It's a side effect of the way many of the AWS APIs trigger a change and return "success" before the change has actually been applied.  Over time we've been adding waiters for the more "unstable" tests, but the API for waiters results in extra API calls, and in turn sometimes causes problems with the AWS rate limiting.

**Comment by yugesk** [comment]
Are we waiting on #1845 for merging this one?

**Comment by tremble** [comment]
recheck

**Comment by tremble** [comment]
@yugesk,

While the integration tests are a little flaky (hopefully #1845 and #1848 will improve this), I don't consider them a blocker.  We can retry the tests without a big problem.

Coming from anyone other than an Amazon developer, I suspect we would have turned down this PR because we can't test it and the vast bulk of folks using this collection can't use the option.  However, your profile seems to show you are an Amazon Developer, so I'm open it its inclusion.  However, while I work for Red Hat, I'm officially a "community contributor" as far as Ansible is concerned and have other responsibilities.  As such I would prefer that someone from the Ansible team approves this PR before it's merged, especially since they're the ones who have to deal with things when a customer opens a case...

File: plugins/modules/ec2_instance.py
Comment (alinabuzachis) [inline plugins/modules/ec2_instance.py:376]: Please add `version_added: 7.1.0`

File: plugins/modules/ec2_instance.py
Comment (alinabuzachis) [inline plugins/modules/ec2_instance.py:955]: Please add `version_added: 7.1.0`

File: plugins/modules/ec2_instance.py
Comment (tremble) [inline plugins/modules/ec2_instance.py:379]:
```suggestion
    type: str
    version_added: 7.1.0
```

File: plugins/modules/ec2_instance.py
Comment (tremble) [inline plugins/modules/ec2_instance.py:959]:
```suggestion
            version_added: 7.1.0
            sample:
```

**Review by tremble — APPROVED** [review]

**Comment by tremble** [comment]
I've squashed the original commits and rebased against main to ensure we pull in the sanity fixes.

In general it's preferable (IMO) to *rebase* rather than merging main into a branch as it keeps the commit history much cleaner.

**Review by alinabuzachis — APPROVED** [review]

**Comment by tremble** [comment]
recheck

**Comment by tremble** [comment]
regate

**Comment by tremble** [comment]
regate

**Pull request merged**
